lineapy/graph_reader/artifact_collection.py

```diff
@@ -110,92 +110,186 @@ def __init__(
+                        slugify(to_artname),
                     )
                     for from_artname in from_artname_set
                 )
                 for to_artname, from_artname_set in dependencies.items()
             )
         )
+
+        self._validate_dependencies_used(inter_session_taskgraph, dependencies)
+
+        inter_session_taskgraph.graph.add_edges_from(task_dependency_edges)
+
+        inter_session_taskgraph = inter_session_taskgraph.remap_nodes(
+            task_name_to_session_id
+        )
+
         # remove loops in graph for dependencies within a session
-        session_id_edges = [
-            edge for edge in session_id_edges if edge[0] != edge[1]
-        ]
-        inter_session_graph = nx.DiGraph()
-        inter_session_graph.add_nodes_from(session_id_nodes)
-        inter_session_graph.add_edges_from(session_id_edges)
-        return inter_session_graph
+        inter_session_taskgraph.remove_self_loops()
+        self._validate_graph_acyclic(inter_session_taskgraph)
+
+        return inter_session_taskgraph
+
+    def create_inter_artifact_taskgraph(
+        self, dependencies: TaskGraphEdge
+    ) -> TaskGraph:
+        """
+        Returns TaskGraph for dependencies between "Artifacts".
+
+        Nodes correspond to nodecollections.
+
+        Edges between artifacts in different sessions are replaced
+        so that all the dependencies pointing to the target artifact instead point to
+        its "root" tasks in the session which are source tasks that can be used to generate the artifact.
+
+        Session 1: A -> B
+        Session 2: C -> D
+
+        If a user specifies a dependency from B -> D, the inter artifact taskgraph will
+        instead draw a dependency from B -> C to ensure B runs before all the intra session
+        dependencies of D.
+
+        Note: These additional edges prevent bugs as the user maybe be unable to specify dependencies
+        to automatically generated components.
+        """
+        inter_artifact_taskgraph = TaskGraph(nodes=[], edges={})
+
+        # add subgraph for each session_artifact
+        for sa in self.session_artifacts.values():
+            inter_artifact_taskgraph.graph = nx.compose(
+                inter_artifact_taskgraph.graph,
+                sa.nodecollection_dependencies.graph,
+            )
+
+        art_name_to_session_id: Dict[str, str] = {}
+        for artifact in self.target_artifacts:
+            art_name_to_session_id[slugify(artifact.name)] = str(
+                artifact._session_id
+            )
+
+        self._validate_dependencies_used(
+            inter_artifact_taskgraph, dependencies
+        )
+
+        for to_artname, from_artname_set in dependencies.items():
+            for from_artname in from_artname_set:
+                from_session_id, to_session_id = (
+                    art_name_to_session_id[from_artname],
+                    art_name_to_session_id[to_artname],
+                )
```

Comment:
Although `dependencies` is currently being passed to `create_inter_artifact_taskgraph()` after being slugified, I am not sure if we can assume this is always the case. That is, given that `create_inter_artifact_taskgraph()` is put as a publicly exposed method, "raw" user-provided `dependencies` may be passed without slugification. To handle such a case, we may want to explicitly slugify `to_artname` and `from_artname` here as well.

Note that `create_inter_session_taskgraph()` is already doing this, like so:

```python
task_dependency_edges = list(
    chain.from_iterable(
        (
            (
                slugify(from_artname),
                slugify(to_artname),
            )
            for from_artname in from_artname_set
        )
        for to_artname, from_artname_set in dependencies.items()
    )
)
```

Reply:
Derp, this is a mistake for sure. 

Adding function `slugify_dependencies` that is pulled from code in the constructor right now and adding it to the top of these functions.

Reply:
Ok, I am understanding that this is the same case for `create_inter_artifact_taskgraph()`, i.e., we don't need to create the inter-artifact task graph again after `ArtifactCollection` is initialized.

In that case, why not make these methods "private" (i.e., underscore-prefixed)? This way, we explicitly signify that only attributes are to be accessed.

Reply:
uuuh yes, i will make them private